servers/zms/src/main/java/com/yahoo/athenz/zms/store/impl/jdbc/JDBCConnection.java

```diff
@@ -150,21 +150,28 @@
             + "(role_id, admin, member, action, audit_ref) VALUES (?,?,?,?,?);";
     private static final String SQL_LIST_ROLE_AUDIT_LOGS = "SELECT * FROM role_audit_log WHERE role_id=?;";
     private static final String SQL_GET_POLICY = "SELECT * FROM policy "
-            + "JOIN domain ON domain.domain_id=policy.domain_id WHERE domain.name=? AND policy.name=?;";
+            + "JOIN domain ON domain.domain_id=policy.domain_id WHERE domain.name=? AND policy.name=? AND policy.active=true;";
+    private static final String SQL_GET_POLICY_VERSION = "SELECT * FROM policy "
+            + "JOIN domain ON domain.domain_id=policy.domain_id WHERE domain.name=? AND policy.name=? AND policy.version=?;";
     private static final String SQL_INSERT_POLICY = "INSERT INTO policy (name, domain_id) VALUES (?,?);";
+    private static final String SQL_INSERT_POLICY_VERSION = "INSERT INTO policy (name, domain_id, version, active) VALUES (?,?, ?, false);";
     private static final String SQL_UPDATE_POLICY = "UPDATE policy SET name=? WHERE policy_id=?;";
     private static final String SQL_UPDATE_POLICY_MOD_TIMESTAMP = "UPDATE policy "
             + "SET modified=CURRENT_TIMESTAMP(3) WHERE policy_id=?;";
-    private static final String SQL_GET_POLICY_ID = "SELECT policy_id FROM policy WHERE domain_id=? AND name=?;";
+    private static final String SQL_SET_ACTIVE_POLICY_VERSION = "UPDATE policy SET active = CASE WHEN version=? then true ELSE false END WHERE domain_id=? AND name=?;";
+    private static final String SQL_GET_ACTIVE_POLICY_ID = "SELECT policy_id FROM policy WHERE domain_id=? AND name=? AND active=true;";
+    private static final String SQL_GET_POLICY_VERSION_ID = "SELECT policy_id FROM policy WHERE domain_id=? AND name=? AND version=?;";
     private static final String SQL_DELETE_POLICY = "DELETE FROM policy WHERE domain_id=? AND name=?;";
+    private static final String SQL_DELETE_POLICY_VERSION = "DELETE FROM policy WHERE domain_id=? AND name=? AND version=? and active=false;";
     private static final String SQL_LIST_POLICY = "SELECT name FROM policy WHERE domain_id=? AND active=1";
+    private static final String SQL_LIST_POLICY_VERSION = "SELECT version FROM policy WHERE domain_id=? AND name=?";
     private static final String SQL_COUNT_POLICY = "SELECT COUNT(*) FROM policy WHERE domain_id=?";
     private static final String SQL_LIST_ASSERTION = "SELECT * FROM assertion WHERE policy_id=?";
     private static final String SQL_COUNT_ASSERTION = "SELECT COUNT(*) FROM assertion WHERE policy_id=?";
     private static final String SQL_GET_ASSERTION = "SELECT * FROM assertion "
             + "JOIN policy ON assertion.policy_id=policy.policy_id "
             + "JOIN domain ON policy.domain_id=domain.domain_id "
-            + "WHERE assertion.assertion_id=? AND domain.name=? AND policy.name=?;";
+            + "WHERE assertion.assertion_id=? AND domain.name=? AND policy.name=? AND policy.active=true;";
```

Comment:
Can we review this usage - this looks suspicious to me. for the given request we already have the assertion_id which means we know exactly what version of the policy this belongs to, so why are we adding an active check?

Looking at this a little bit more - this seems to be used in the delete assertion api call from zmsimpl/dbservice so how are we deleting assertions from non-active versions of the policy

Reply:
Well at the moment we don't. I assumed that we can only modify active policy versions as the new endpoints didn't mention providing the option to add / remove assertions for non-active versions. So if we want to add or remove assertions we need to make the policy active first.
I understand now that this should be supported and will update the API accordingly

Reply:
Fixed (added 2 new endpoints to delete / put assertions for policy versions.